File: apps/android/app/src/main/java/com/farcooler/model/DiffComputation.kt
package com.farcooler.model

// The diff a tool call's edit is rendered from.
//
// Ported from `apps/shared/AgentKit/Sources/AgentKit/DiffComputation.swift`,
// which exists because the algorithm was written once and then copied — same
// trimming, same LCS, same 160,000-cell cutoff — with only the doc comment
// drifting between the two Apple targets. Two copies of a diff algorithm is two
// places for an off-by-one to live and one place for it to be fixed. There is
// no way to share Swift with Kotlin, so this is a third copy; the unit tests
// are what hold it to the same answers.

/** The diffing itself, kept apart from the view so it has no Compose in it. */
object DiffComputation {
    enum class Kind { CONTEXT, ADDED, REMOVED }

    data class Line(
        val id: Int,
        val kind: Kind,
        val oldNumber: Int?,
        val newNumber: Int?,
        val text: String,
    )

    /**
     * Line-based diff of two whole file texts.
     *
     * A real Myers diff is O(n·d) and would be the right tool against two
     * arbitrary files; what a tool call actually hands over is usually one small
     * change inside two mostly-identical texts, so trimming the common prefix
     * and suffix first — O(n) — throws away nearly all of the work before the
     * expensive part ever runs. The expensive part, an LCS over whatever is left
     * in the middle, is only worth its O(n·m) on the leftover sliver; past a
     * size where that would be slow it falls back to "the whole middle changed",
     * which is still a correct diff, just not the minimal one.
     */
    fun compute(old: String, new: String): List<Line> {
        val oldLines = if (old.isEmpty()) emptyList() else old.split("\n")
        val newLines = if (new.isEmpty()) emptyList() else new.split("\n")

        var prefix = 0
        while (prefix < oldLines.size && prefix < newLines.size &&
            oldLines[prefix] == newLines[prefix]
        ) {
            prefix += 1
        }

        var suffix = 0
        while (suffix < oldLines.size - prefix && suffix < newLines.size - prefix &&
            oldLines[oldLines.size - 1 - suffix] == newLines[newLines.size - 1 - suffix]
        ) {
            suffix += 1
        }

        val oldMiddle = oldLines.subList(prefix, oldLines.size - suffix)
        val newMiddle = newLines.subList(prefix, newLines.size - suffix)

        val result = mutableListOf<Line>()
        var id = 0
        var oldNum = 1
        var newNum = 1

        fun push(kind: Kind, oldNumber: Int?, newNumber: Int?, text: String) {
            result.add(Line(id, kind, oldNumber, newNumber, text))
            id += 1
        }

        for (line in oldLines.subList(0, prefix)) {
            push(Kind.CONTEXT, oldNum, newNum, line)
            oldNum += 1
            newNum += 1
        }

        // 400*400 is a worst case of 160,000 cells, which is fast; past that
        // the middle is shown as a flat replace rather than paying for the
        // table.
        if (oldMiddle.size * newMiddle.size <= 160_000) {
            for (op in lcsDiff(oldMiddle, newMiddle)) {
                when (op) {
                    is Op.Equal -> {
                        push(Kind.CONTEXT, oldNum, newNum, op.line)
                        oldNum += 1
                        newNum += 1
                    }

                    is Op.Removed -> {
                        push(Kind.REMOVED, oldNum, null, op.line)
                        oldNum += 1
                    }

                    is Op.Added -> {
                        push(Kind.ADDED, null, newNum, op.line)
                        newNum += 1
                    }
                }
            }
        } else {
            for (line in oldMiddle) {
                push(Kind.REMOVED, oldNum, null, line)
                oldNum += 1
            }
            for (line in newMiddle) {
                push(Kind.ADDED, null, newNum, line)
                newNum += 1
            }
        }

        val suffixStart = oldLines.size - suffix
        for (offset in 0 until suffix) {
            push(Kind.CONTEXT, oldNum, newNum, oldLines[suffixStart + offset])
            oldNum += 1
            newNum += 1
        }

        return result
    }

    private sealed interface Op {
        data class Equal(val line: String) : Op
        data class Removed(val line: String) : Op
        data class Added(val line: String) : Op
    }

    /** Classic LCS table, walked backwards to recover the edit script. */
    private fun lcsDiff(a: List<String>, b: List<String>): List<Op> {
        if (a.isEmpty()) return b.map { Op.Added(it) }
        if (b.isEmpty()) return a.map { Op.Removed(it) }

        val table = Array(a.size + 1) { IntArray(b.size + 1) }
        for (i in a.indices.reversed()) {
            for (j in b.indices.reversed()) {
                table[i][j] =
                    if (a[i] == b[j]) table[i + 1][j + 1] + 1
                    else maxOf(table[i + 1][j], table[i][j + 1])
            }
        }

        val ops = mutableListOf<Op>()
        var i = 0
        var j = 0
        while (i < a.size && j < b.size) {
            when {
                a[i] == b[j] -> {
                    ops.add(Op.Equal(a[i]))
                    i += 1
                    j += 1
                }

                table[i + 1][j] >= table[i][j + 1] -> {
                    ops.add(Op.Removed(a[i]))
                    i += 1
                }

                else -> {
                    ops.add(Op.Added(b[j]))
                    j += 1
                }
            }
        }
        while (i < a.size) {
            ops.add(Op.Removed(a[i]))
            i += 1
        }
        while (j < b.size) {
            ops.add(Op.Added(b[j]))
            j += 1
        }
        return ops
    }
}

File: apps/shared/AgentKit/Sources/AgentKit/DiffComputation.swift
import Foundation

// The diff a tool call's edit is rendered from, shared by both apps.
//
// Pure Foundation: no SwiftUI, no AppKit, no UIKit. It was written once and
// then copied, so the two targets carried the same algorithm — the same
// trimming, the same LCS, the same 160,000-cell cutoff — with only the doc
// comment drifting between them. Two copies of a diff algorithm is two places
// for an off-by-one to live and one place for it to be fixed.

/// The diffing itself, kept apart from the view so it has no SwiftUI in it.
public enum DiffComputation {
    public enum Kind { case context, added, removed }

    public struct Line: Identifiable {
        public let id: Int
        public let kind: Kind
        public let oldNumber: Int?
        public let newNumber: Int?
        public let text: String

        /// Public so the review surface renders the SAME line model the agent
        /// transcript does, rather than declaring a second one that is free to
        /// disagree about what a changed line is.
        public init(id: Int, kind: Kind, oldNumber: Int?, newNumber: Int?, text: String) {
            self.id = id
            self.kind = kind
            self.oldNumber = oldNumber
            self.newNumber = newNumber
            self.text = text
        }
    }

    /// Line-based diff of two whole file texts.
    ///
    /// A real Myers diff is O(n·d) and would be the right tool against two
    /// arbitrary files; what a tool call actually hands over is usually one
    /// small change inside two mostly-identical texts, so trimming the common
    /// prefix and suffix first — O(n) — throws away nearly all of the work
    /// before the expensive part ever runs. The expensive part, an LCS over
    /// whatever is left in the middle, is only worth its O(n·m) on the
    /// leftover sliver; past a size where that would be slow it falls back to
    /// "the whole middle changed", which is still a correct diff, just not the
    /// minimal one.
    public static func compute(old: String, new: String) -> [Line] {
        let oldLines = old.isEmpty ? [] : old.components(separatedBy: "\n")
        let newLines = new.isEmpty ? [] : new.components(separatedBy: "\n")

        var prefix = 0
        while prefix < oldLines.count, prefix < newLines.count,
            oldLines[prefix] == newLines[prefix]
        {
            prefix += 1
        }

        var suffix = 0
        while suffix < oldLines.count - prefix, suffix < newLines.count - prefix,
            oldLines[oldLines.count - 1 - suffix] == newLines[newLines.count - 1 - suffix]
        {
            suffix += 1
        }

        let oldMiddle = Array(oldLines[prefix..<(oldLines.count - suffix)])
        let newMiddle = Array(newLines[prefix..<(newLines.count - suffix)])

        var result: [Line] = []
        var id = 0
        var oldNum = 1
        var newNum = 1

        func push(_ kind: Kind, old: Int?, new: Int?, text: String) {
            result.append(Line(id: id, kind: kind, oldNumber: old, newNumber: new, text: text))
            id += 1
        }

        for line in oldLines[0..<prefix] {
            push(.context, old: oldNum, new: newNum, text: line)
            oldNum += 1
            newNum += 1
        }

        // 400*400 is a worst case of 160,000 cells, which is fast; past that
        // the middle is shown as a flat replace rather than paying for the
        // table.
        if oldMiddle.count * newMiddle.count <= 160_000 {
            for op in lcsDiff(oldMiddle, newMiddle) {
                switch op {
                case let .equal(line):
                    push(.context, old: oldNum, new: newNum, text: line)
                    oldNum += 1
                    newNum += 1
                case let .removed(line):
                    push(.removed, old: oldNum, new: nil, text: line)
                    oldNum += 1
                case let .added(line):
                    push(.added, old: nil, new: newNum, text: line)
                    newNum += 1
                }
            }
        } else {
            for line in oldMiddle {
                push(.removed, old: oldNum, new: nil, text: line)
                oldNum += 1
            }
            for line in newMiddle {
                push(.added, old: nil, new: newNum, text: line)
                newNum += 1
            }
        }

        let suffixStart = oldLines.count - suffix
        for offset in 0..<suffix {
            push(.context, old: oldNum, new: newNum, text: oldLines[suffixStart + offset])
            oldNum += 1
            newNum += 1
        }

        return result
    }

    private enum Op {
        case equal(String)
        case removed(String)
        case added(String)
    }

    /// Classic LCS table, walked backwards to recover the edit script.
    private static func lcsDiff(_ a: [String], _ b: [String]) -> [Op] {
        guard !a.isEmpty else { return b.map { .added($0) } }
        guard !b.isEmpty else { return a.map { .removed($0) } }

        var table = [[Int]](repeating: [Int](repeating: 0, count: b.count + 1), count: a.count + 1)
        for i in stride(from: a.count - 1, through: 0, by: -1) {
            for j in stride(from: b.count - 1, through: 0, by: -1) {
                table[i][j] =
                    a[i] == b[j]
                    ? table[i + 1][j + 1] + 1
                    : max(table[i + 1][j], table[i][j + 1])
            }
        }

        var ops: [Op] = []
        var i = 0, j = 0
        while i < a.count && j < b.count {
            if a[i] == b[j] {
                ops.append(.equal(a[i]))
                i += 1
                j += 1
            } else if table[i + 1][j] >= table[i][j + 1] {
                ops.append(.removed(a[i]))
                i += 1
            } else {
                ops.append(.added(b[j]))
                j += 1
            }
        }
        while i < a.count { ops.append(.removed(a[i])); i += 1 }
        while j < b.count { ops.append(.added(b[j])); j += 1 }
        return ops
    }
}
